File: ios/iosApp/Views/Example/ExampleView.swift
import Shared
import SwiftUI

struct ExampleView: View {

    private let viewModel: ExampleViewModel

    @State private var state: ExampleViewModelState

    init(viewModel: ExampleViewModel) {
        self.viewModel = viewModel
        self.state = viewModel.state.value
    }

	var body: some View {
        VStack {
            Text(state.text)

            LazyVStack {
                ForEach(state.items, id: \.identifier) { item in
                    Text(item.text)
                }
            }

            ButtonView(viewModel.button)

            ToggleView(viewModel.toggle)

            ImageView(viewModel.image)
                .aspectRatio(contentMode: .fit)
        }
        .padding()
        .task {
            for await state in viewModel.state {
                self.state = state
            }
        }
    }
}

#Preview {
    PreviewProvider { viewModelFactory in
        ExampleView(viewModel: viewModelFactory.exampleViewModel())
    }
}

File: androidApp/src/main/java/com/mirego/kmp/boilerplate/views/example/ExampleView.kt
package com.mirego.kmp.boilerplate.views.example

import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.ui.tooling.preview.Preview
import com.mirego.kmp.boilerplate.previews.PreviewProvider
import com.mirego.kmp.boilerplate.viewmodels.example.ExampleViewModel
import com.mirego.kmp.boilerplate.views.components.ButtonView
import com.mirego.kmp.boilerplate.views.components.ImageView
import com.mirego.kmp.boilerplate.views.components.ToggleView

@Composable
fun ExampleView(viewModel: ExampleViewModel) {
    val state: ExampleViewModel.State by viewModel.state.collectAsState()

    Column {
        Text(state.text)

        LazyColumn {
            items(state.items, key = ExampleViewModel.State.Item::identifier) { item ->
                Text(text = item.text)
            }
        }

        ButtonView(viewModel = viewModel.button)

        ToggleView(viewModel = viewModel.toggle)

        ImageView(viewModel = viewModel.image)
    }
}

@Preview(showSystemUi = true)
@Composable
fun PreviewGreeting() {
    PreviewProvider { coroutineScope, viewModelFactory ->
        ExampleView(viewModel = viewModelFactory.exampleViewModel(coroutineScope))
    }
}
